ilastik/applets/dataSelection/opDataSelection.py

```diff
@@ -1016,6 +1058,12 @@ def propagateDirty(self, slot, subindex, roi):
         # Output slots are directly connected to internal operators
         pass
 
+    def _ensure_transaction_in_progress(self, *_):
+        # Setting scale while the op is fully ready is forbidden.
+        # Doing this will set scale on the internal OpDataSelection for each data role one by one,
+        # causing setupOutputs throughout the workflow with different data roles providing different data shape.
+        assert not self.ScaleChangeFinished.ready(), "must disconnect ScaleChangeFinished first before changing scale"
```

Comment:
I like the pattern, but if you want to rely on it (and also I've seen you test this behavior) it should probably not be an assert, but raise a dedicated exception?